File: Trees/100_Same_Tree.py
# https://leetcode.com/problems/same-tree/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        # base case 1 - if both don't exist, match!
        if not p and not q:
            return True

        # base case 2 - if one don't exist, not match!
        if not p or not q:
            return False

        # baes case 3 - if they are not equal, not match!
        if p.val != q.val:
            return False
        
        # return answer (both need to be true!)
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)

File: Trees/100_Same_Tree.js
// https://leetcode.com/problems/same-tree
// TIME COMPLEXITY: O(n)
// SPACE COMPLEXITY: O(h) h is height of tree due to recursion stack

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {boolean}
 */
var isSameTree = function(p, q) {
    const dfs = (one, two) => {
        // base cases: if end of branch and if not equal
        if (!one && !two) return true;
        if (!one || !two || one.val !== two.val) return false;
        return (dfs(one.left, two.left) && dfs(one.right, two.right))
        
        // previous logic - bad because trying to check children! instead trust base case and pass children to the next recurse
        // if (one.val !== two.val) return false;
        // if (one.left && !two.left) return false;
        // ...
    }
    return dfs (p, q);
};